main.vala

```diff
@@ -1,68 +1,64 @@
-    private string filename;
-
-    public Vala.SourceFile file;
-    public string uri;
-    public int version;
-
-    public TextDocument (Context ctx,
-                         string filename,
-                         string? content = null,
-                         int version = 0) throws ConvertError, FileError {
-
-        if (!FileUtils.test (filename, FileTest.EXISTS)) {
-            throw new FileError.NOENT ("file %s does not exist".printf (filename));
-        }
+class Vls.Request {
+    private int64? int_value;
+    private string? string_value;
+    private string? method;
+
+    public Request (Variant id, string? method = null) {
+        assert (id.is_of_type (VariantType.INT64) || id.is_of_type (VariantType.STRING));
+        if (id.is_of_type (VariantType.INT64))
+            int_value = (int64) id;
+        else
+            string_value = (string) id;
+        this.method = method;
+    }
 
-        this.uri = Filename.to_uri (filename);
-        this.filename = filename;
-        this.version = version;
-        this.ctx = ctx;
+    public string to_string () {
+        string id_string = int_value != null ? int_value.to_string () : string_value;
+        return id_string + (method != null ? @":$method" : "");
+    }
 
-        var type = Vala.SourceFileType.NONE;
-        if (uri.has_suffix (".vala") || uri.has_suffix (".gs"))
-            type = Vala.SourceFileType.SOURCE;
-        else if (uri.has_suffix (".vapi"))
-            type = Vala.SourceFileType.PACKAGE;
+    public static uint hash (Request req) {
+        if (req.int_value != null)
+            return GLib.int64_hash (req.int_value);
+        else
+            return GLib.str_hash (req.string_value);
+    }
 
-        file = new Vala.SourceFile (ctx.code_context, type, filename, content);
-        if (type == Vala.SourceFileType.SOURCE) {
-            var ns_ref = new Vala.UsingDirective (new Vala.UnresolvedSymbol (null, "GLib", null));
-            file.add_using_directive (ns_ref);
-            ctx.add_using ("GLib");
+    public static bool equal (Request reqA, Request reqB) {
+        if (reqA.int_value != null) {
+            assert (reqB.int_value != null);
+            return reqA.int_value == reqB.int_value;
+        } else {
+            assert (reqB.string_value != null);
+            return reqA.string_value == reqB.string_value;
         }
     }
 }
 
+errordomain Vls.ProjectError {
+    INTROSPECT,
+    JSON
+}
+
 class Vls.Server {
     Jsonrpc.Server server;
     MainLoop loop;
-    HashTable<string, string> cc;
-    Context ctx;
+
     HashTable<string, NotificationHandler> notif_handlers;
     HashTable<string, CallHandler> call_handlers;
     InitializeParams init_params;
 
+    const uint check_update_context_period_ms = 200;
+    const int64 update_context_delay_inc_us = check_update_context_period_ms * 50;
+    const int64 update_context_delay_max_us = 1000 * 1000;
+    const uint wait_for_context_update_delay_ms = 500;
```

Comment:
This is for performance reasons with nontrivially-sized projects. The old way (rebuilding on demand) would slow to a crawl when typing. It doesn't make sense to recompile on every keystroke since your next keystroke invalidates all of your work, and so you spend a lot of time compiling meanwhile your next server message comes in and you can't respond to it because you're still compiling...

It's much faster to wait for the user to "finish" typing and then compile everything.